modules/SerialCommHub/tiny_modbus_rtu.cpp

```diff
@@ -103,92 +134,112 @@ static std::vector<uint16_t> decode_reply(const uint8_t* buf, int len, uint8_t e
             break;
         case 0x03:
-            EVLOG_error << "Modbus exception: Illegal data value";
+            throw ModbusException("Modbus exception: Illegal data value");
             break;
         case 0x04:
-            EVLOG_error << "Modbus exception: Client device failure";
+            throw ModbusException("Modbus exception: Client device failure");
             break;
         case 0x05:
-            EVLOG_debug << "Modbus ACK";
+            throw ModbusException("Modbus ACK");
             break;
         case 0x06:
-            EVLOG_error << "Modbus exception: Client device busy";
+            throw ModbusException("Modbus exception: Client device busy");
             break;
         case 0x07:
-            EVLOG_error << "Modbus exception: NACK";
+            throw ModbusException("Modbus exception: NACK");
             break;
         case 0x08:
-            EVLOG_error << "Modbus exception: Memory parity error";
+            throw ModbusException("Modbus exception: Memory parity error");
             break;
         case 0x09:
-            EVLOG_error << "Modbus exception: Out of resources";
+            throw ModbusException("Modbus exception: Out of resources");
             break;
         case 0x0A:
-            EVLOG_error << "Modbus exception: Gateway path unavailable";
+            throw ModbusException("Modbus exception: Gateway path unavailable");
             break;
         case 0x0B:
-            EVLOG_error << "Modbus exception: Gateway target device failed to respond";
+            throw ModbusException("Modbus exception: Gateway target device failed to respond");
             break;
         default:
-            EVLOG_error << "Modbus exception: Unknown";
+            throw ModbusException("Modbus exception: Unknown");
         }
-        return result;
     }
+
+    // For a write reply we always get 4 bytes
+    uint8_t byte_cnt = 4;
+    int start_of_result = RES_TX_START_OF_PAYLOAD;
+    bool even_byte_cnt_expected = false;
+
+    // Was it a read reply?
+    switch (function) {
+    case FunctionCode::WRITE_SINGLE_COIL:
+    case FunctionCode::WRITE_SINGLE_HOLDING_REGISTER:
+    case FunctionCode::WRITE_MULTIPLE_COILS:
+    case FunctionCode::WRITE_MULTIPLE_HOLDING_REGISTERS:
+        // no - nothing to do
+        break;
+    case FunctionCode::READ_MULTIPLE_HOLDING_REGISTERS:
+    case FunctionCode::READ_INPUT_REGISTERS:
+        // yes - for 16-bit wide registers thus we can assume a even byte count
+        even_byte_cnt_expected = true;
+        [[fallthrough]];
+    case FunctionCode::READ_COILS:
+    case FunctionCode::READ_DISCRETE_INPUTS:
+        // yes
+        // adapt byte count and starting pos
+        byte_cnt = buf[RES_RX_LEN_POS];
+        start_of_result = RES_RX_START_OF_PAYLOAD;
+        break;
+    default:
+        throw std::logic_error("Missing implementation for function code " + FunctionCode_to_string_with_hex(function));
```

Comment:
This exception will not be caught as far as I can see. We are trying to get rid of all uncaught exception throughout the code base (as well as all exit() from modules in case of error). Even if it is an error in the code if the implementation is missing here, it should not trigger a complete restart of the system including ocpp disconnection. We should just indicate that the modbus transfer failed, as we do with other transport errors and not exit the process.

Reply:
Will fix.

So we should also avoid the other 
```throw std::system_error(errno, std::generic_category(), "Could not send Modbus request");```
when `write()` fails, right?

Reply:
yes, see my comment further down

Reply:
Is this still relevant? As far as I can see this wasn't addressed

Reply:
same here